internal/auth/flow.go

```diff
@@ -0,0 +1,226 @@
+		return nil, fmt.Errorf("authentication timeout after 5 minutes")
+	}
+}
+
+func (f *Flow) buildAuthURL() string {
+	redirectURI := fmt.Sprintf("http://localhost:%d%s", f.RedirectPort, CallbackPath)
+	codeChallenge := generateCodeChallenge(f.codeVerifier)
+
+	params := url.Values{}
+	params.Set("client_id", f.ClientID)
+	params.Set("response_type", "code")
+	params.Set("redirect_uri", redirectURI)
+	params.Set("state", f.state)
+	params.Set("code_challenge", codeChallenge)
+	params.Set("code_challenge_method", "S256")
+	params.Set("scope", "openid profile email offline_access")
+
+	return fmt.Sprintf("%s?%s", f.Provider.AuthorizationEndpoint, params.Encode())
+}
+
+func (f *Flow) handleCallback(listener net.Listener, codeChan chan<- string, errChan chan<- error) {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(CallbackPath, func(w http.ResponseWriter, r *http.Request) {
+		// Validate state
+		if state := r.URL.Query().Get("state"); state != f.state {
+			errChan <- fmt.Errorf("invalid state parameter")
+			http.Error(w, "Invalid state", http.StatusBadRequest)
+			return
+		}
+
+		// Check for errors
+		if errParam := r.URL.Query().Get("error"); errParam != "" {
+			errDesc := r.URL.Query().Get("error_description")
+			errChan <- fmt.Errorf("authentication error: %s - %s", errParam, errDesc)
+			http.Error(w, "Authentication failed", http.StatusBadRequest)
+			return
+		}
+
+		// Get authorization code
+		code := r.URL.Query().Get("code")
+		if code == "" {
+			errChan <- fmt.Errorf("no authorization code received")
+			http.Error(w, "No code received", http.StatusBadRequest)
+			return
+		}
+
+		// Send success response
+		w.Header().Set("Content-Type", "text/html")
+		fmt.Fprint(w, successHTML)
+
+		// Send code
+		codeChan <- code
+	})
+
+	server := &http.Server{Handler: mux}
+	server.Serve(listener)
+}
+
+func (f *Flow) exchangeCode(ctx context.Context, code string) (*Credentials, error) {
+	redirectURI := fmt.Sprintf("http://localhost:%d%s", f.RedirectPort, CallbackPath)
+
+	tokens, err := ExchangeCodeForTokens(ctx, f.Provider, f.ClientID, code, redirectURI, f.codeVerifier)
+	if err != nil {
+		return nil, fmt.Errorf("failed to exchange code for tokens: %w", err)
+	}
+
+	// Calculate expiration time
+	expiresAt := time.Now()
+	if tokens.ExpiresIn > 0 {
+		expiresAt = expiresAt.Add(time.Duration(tokens.ExpiresIn) * time.Second)
+	} else {
+		// Default to 1 hour if not specified
+		expiresAt = expiresAt.Add(time.Hour)
+	}
+
+	return &Credentials{
+		AccessToken:  tokens.AccessToken,
+		RefreshToken: tokens.RefreshToken,
+		TokenType:    tokens.TokenType,
+		ExpiresAt:    expiresAt,
+	}, nil
+}
+
+// PKCE helpers
+
+func generateCodeVerifier() string {
+	b := make([]byte, 32)
+	rand.Read(b)
+	return base64.RawURLEncoding.EncodeToString(b)
+}
+
+func generateCodeChallenge(verifier string) string {
+	hash := sha256.Sum256([]byte(verifier))
+	return base64.RawURLEncoding.EncodeToString(hash[:])
+}
+
+const successHTML = `<!DOCTYPE html>
```

Comment:
For code-cleanliness reasons I wonder if we should move this to a separate file and then embed it into this file.